core/server/common/src/main/java/alluxio/master/journal/FileSystemMergeJournalContext.java

```diff
@@ -0,0 +1,110 @@
+ */
+
+package alluxio.master.journal;
+
+import alluxio.conf.Configuration;
+import alluxio.conf.PropertyKey;
+import alluxio.exception.status.UnavailableException;
+import alluxio.proto.journal.Journal.JournalEntry;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Context for merging journal entries together for a wrapped journal context.
+ *
+ * This is used so that we can combine several journal entries into one using a merge
+ * function. This helps us resolve the inconsistency between primary and standby and
+ * decrease the chance for a standby to see intermediate state of a file system operation.
+ */
+@NotThreadSafe
+public final class FileSystemMergeJournalContext implements JournalContext {
+  // It will log a warning if the number of buffered journal entries exceed 100
+  public static final int MAX_ENTRIES
+      = Configuration.getInt(PropertyKey.MASTER_MERGE_JOURNAL_CONTEXT_MAX_ENTRIES);
+
+  private static final Logger LOG = LoggerFactory.getLogger(MergeJournalContext.class);
+
+  private final JournalContext mJournalContext;
+  private final JournalEntryMerger mJournalEntryMerger;
+
+  /**
+   * Constructs a {@link FileSystemMergeJournalContext}.
+   * @param journalContext the journal context to wrap
+   * @param journalEntryMerger the merger which merges multiple journal entries into one
+   */
+  public FileSystemMergeJournalContext(JournalContext journalContext,
+                             JournalEntryMerger journalEntryMerger) {
+    Preconditions.checkNotNull(journalContext, "journalContext is null");
+    Preconditions.checkNotNull(journalEntryMerger, "mergeOperator is null");
+    mJournalContext = journalContext;
+    mJournalEntryMerger = journalEntryMerger;
+  }
+
+  /**
+   * Adds the new journal entry into the journal entry merger.
+   * If the size of the outstanding journal entries are bigger than the threshold,
+   * we will flush them immediately into the async journal writer.
+   * This is used to prevent a recursive file system operation adding
+   * too many journal entries without being flushed and taking huge memory.
+   * Though it might create inconsistency between primary and standby if
+   * the primary crashes between two flushes.
+   * Tune PropertyKey.MASTER_MERGE_JOURNAL_CONTEXT_MAX_ENTRIES to trade off
+   * @param entry the {@link JournalEntry} to append to the journal
+   */
+  @Override
+  public void append(JournalEntry entry) {
+    mJournalEntryMerger.add(entry);
+    List<JournalEntry> journalEntries = mJournalEntryMerger.getMergedJournalEntries();
+    if (journalEntries.size() >= MAX_ENTRIES) {
+      LOG.warn("MergeJournalContext has " + journalEntries.size()
+          + " entries, over the limit of " + MAX_ENTRIES + ", triggering force flushing");
+      appendMergedJournals();
```

Comment:
force flush journals here to address your comment @tcrain

Reply:
The atomicity of operation may be broken here (as broken as the current alluxio). Let's leave a github issue link here.

Reply:
Yes, I would rather not do it here, and give the responsibility to the operation owner to manually flush the journals as doing it here could make an operation partially visible. Since only the owning operation knows when it is safe to flush the journal it should be the one who has to do this.

Note that the worry is not about making a big recursive operation partially visible and instead about making individual operations partially visible, for example a single create operation. So flushing at a safe point in a recursive operation should be fine. Most file systems, eg POSIX, do not ensure recursive operations are atomic anyway.

For example within the recursive delete operation, it may keep track of how many items it has deleted so far, then flush the journal manually once a certain threshold has been reached. This way you avoid possibly showing a partially item.

Reply:
okay I made the flush process happen in FSM. @tcrain can you take a look? Thanks!